### PR TITLE
Handle data sources and advanced settings as global object.

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description

Handle data source and config as global objects.
- Listing data source from global
- Create data source into global
- When import, gives error if user tries to import global objects into a workspace.

### Import data source
image

### Copy data source
image

### Create data source in workspace(in normal flow, user won't be able to visit data source management page inside workspace)
image

### Create advanced settings in workspace(in normal flow, user won't be able to visit advanced settings page inside workspace)
image

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/workspace/server/saved_objects/integration_tests/workspace_id_consumer_wrapper.test.ts [hunk 1]
```diff
@@ -140,6 +159,79 @@ describe('workspace_id_consumer integration test', () => {
...
+        })
+      );
+      expect(createResultFoo.body.saved_objects[1].error).toEqual(
+        expect.objectContaining({
+          message: "Unsupport type in workspace: 'config' is not allowed to import in workspace.",
+          statusCode: 400,
+        })
+      );
+
+      // data source and advanced settings should not be found within the workspace
```
Comment (Hailong-am) [inline src/plugins/workspace/server/saved_objects/integration_tests/workspace_id_consumer_wrapper.test.ts:193]: Does it means when create index pattern, there is no way to select datasource?

Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/integration_tests/workspace_id_consumer_wrapper.test.ts:193]: This comment means the data source should not be created inside the workspace and thus can not be found. Let me adjust this comment a little bit.

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts [hunk 2]
```diff
@@ -46,14 +68,61 @@ export class WorkspaceIdConsumerWrapper {
           attributes,
           this.formatWorkspaceIdParams(wrapperOptions.request, options)
         ),
-      bulkCreate: <T = unknown>(
+      bulkCreate: async <T = unknown>(
```
Comment (wanglam) [inline src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts:71]: Seems these saved object type check was only exists in bulkCreate, shall we need add similar logic to create method?

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts [hunk 3]
```diff
@@ -37,6 +41,24 @@ export class WorkspaceIdConsumerWrapper {
       ...(finalWorkspaces.length ? { workspaces: finalWorkspaces } : {}),
     };
   }
+  private isDataSourceType(type: SavedObjectsFindOptions['type']): boolean {
```
Comment (raintygao) [inline src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts:44]: The type could also be the object type of SavedObjectsBulkCreateObject ? In this case, type will only be single string.

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts [hunk 4]
```diff
@@ -37,6 +41,24 @@ export class WorkspaceIdConsumerWrapper {
...
+    if (Array.isArray(type)) {
+      return type.every((item) => item === DATA_SOURCE_SAVED_OBJECT_TYPE);
+    }
+
+    return type === DATA_SOURCE_SAVED_OBJECT_TYPE;
+  }
+  private isConfigType(type: SavedObjectsFindOptions['type']): boolean {
```
Comment (raintygao) [inline src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts:51]: Type here should be `SavedObjectsBulkCreateObject['type']` ?

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts (hunk 2 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts:82]: Good point, though user won't be able to visit data source management page in browser but they may curl the OSD API to create disallowed objects. Add that.

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts (hunk 3 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts:44]: The `isDataSourceType` is used within formatFindParams, whose `type` may be string[].

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts (hunk 4 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts]: Sure, changed to `SavedObject['type']`.

File: src/plugins/workspace/server/saved_objects/integration_tests/workspace_id_consumer_wrapper.test.ts [hunk 5]
```diff
@@ -110,6 +129,38 @@ describe('workspace_id_consumer integration test', () => {
...
+          "error": "Bad Request",
+          "message": "Unsupport type in workspace: 'data-source' is not allowed to create in workspace.",
+          "statusCode": 400,
+        }
+      `);
+
+      const createConfigResult = await osdTestServer.request
+        .post(root, `/w/${createdFooWorkspace.id}/api/saved_objects/config`)
+        .send({
+          attributes: dataSource.attributes,
```
Comment (wanglam) [inline src/plugins/workspace/server/saved_objects/integration_tests/workspace_id_consumer_wrapper.test.ts:151]: Shall we use `advancedSettings.attributes`?

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts [hunk 6]
```diff
@@ -37,23 +42,95 @@ export class WorkspaceIdConsumerWrapper {
...
+          saved_objects: [
+            ...allowedSavedObjectsBulkCreateResult.saved_objects,
+            ...disallowedSavedObjects.map((item) => ({
+              references: [],
+              id: '',
+              ...item,
+              error: {
+                ...SavedObjectsErrorHelpers.decorateBadRequestError(
+                  new Error(`'${item.type}' is not allowed to import in workspace.`),
+                  'Unsupport type in workspace'
```
Comment (ruanyl) [inline src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts:126]: `Unsupport` -> `Unsupported`

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts [hunk 7]
```diff
@@ -64,7 +141,14 @@ export class WorkspaceIdConsumerWrapper {
         ),
       delete: wrapperOptions.client.delete,
       find: (options: SavedObjectsFindOptions) =>
-        wrapperOptions.client.find(this.formatWorkspaceIdParams(wrapperOptions.request, options)),
+        wrapperOptions.client.find(
+          this.formatWorkspaceIdParams(
+            wrapperOptions.request,
+            // The `formatFindParams` is a workaroud for 2.14 to always list global data sources,
```
Comment (ruanyl) [inline src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts:147]: `workaroud` -> `workaround`

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts [hunk 8]
```diff
@@ -64,7 +141,14 @@ export class WorkspaceIdConsumerWrapper {
...
+            // should remove this workaround in 2.15 once readonly share is available
```
Comment (ruanyl) [inline src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts:148]: Let's say `in the upcoming release` which doesn't sound too arbitrary. And we can add `TODO: ` in front of the comment.

**Review by ruanyl — COMMENTED** [review]
Does it sound more logical if we put these changes to the workspace conflict check wrapper?

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts (hunk 8 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts]: Sure, done for that.

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts (hunk 7 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts]: Thanks for catching this, done for that.

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts (hunk 6 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts]: Done for that.

File: src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.test.ts [hunk 9]
```diff
@@ -85,6 +118,37 @@ describe('WorkspaceIdConsumerWrapper', () => {
         }
       );
     });
+
+    it(`Should return error when trying to create unallowed type within a workspace`, async () => {
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.test.ts]: Done for that.

**Review by ruanyl — APPROVED** [review]

**Review by raintygao — APPROVED** [review]

**Pull request merged**